### PR TITLE
Improve JSDoc comments

### PR DESCRIPTION
**Pull request by Shinigami92**
Improve some JSDoc comments

like proposed by https://github.com/pioardi/poolifier/pull/116#discussion_r574740337

### CODE REVIEW

**Comment by Shinigami92** [comment]
Hope we get #129 working one day, because it's really annoying that I sometimes forget to run the script in terminal, because `formatOnSave` from VSCode doesn't respect the `prettierx` `spaceBeforeFunctionParen` rule :disappointed:

**Review by pioardi — DISMISSED** [review]
Approved

**Comment by Shinigami92** [comment]
@pioardi Before we just merge this, do you have any other lines, classes, interfaces, types ... where we should improve the JSDocs?  
As this PR is currently in draft-state it's WIP/in progress  
I will look over the whole code today and try to improve further lines

Currently I'm on some other tasks (adSoul work & @prettier/plugin-pug _which is my own main open source project_)

**Comment by pioardi** [comment]
Yes so I think we could add JSDoc to any exposed method and class to explain for what is used, the parameters needed( with a small example if we feel is needed ).

So yes there are further things that I would like to improve on JSDoc in particular because our project is now in TS and also because we could choose to publish the whole API Doc to a github page later on.

File: src/pools/abstract-pool.ts
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:83]: Should we rename this property?

File: src/pools/abstract-pool.ts
Comment (pioardi) [reply src/pools/abstract-pool.ts:250]: Sounds good

Comment (Shinigami92) [reply src/pools/abstract-pool.ts:250]: I will note all the potential renamings and create a new PR for that later
Want to have this a clean JSDoc update PR

File: src/worker/abstract-worker.ts
Comment (pioardi) [inline src/worker/abstract-worker.ts:92]: We only do that for the cluster worker right ?

Comment (Shinigami92) [reply src/worker/abstract-worker.ts:92]: We override it only in the cluster pool.
The thread pool uses the default impl as can be seen here.

The JSDoc lies a bit :sweat_smile:  
Because here we could receive an `Error` and just lie to TS and tell TS that the `Error` is a `string`
You can do that with `(something as unknown) as OtherType`
That is a workaround to tell TS to shut the **** up :rofl:

**Comment by Shinigami92** [comment]
@ST-DDT if you want, you could also have a look  
Your feedback would be helpful because you don't know anything directly of this project  
So you are non-biased and don't have _knowledge_ about the project itself and read it like a newcomer

File: src/pools/abstract-pool.ts
Comment (ST-DDT) [inline src/pools/abstract-pool.ts:21]:
```suggestion
 * Basic interface that describes the minimum required implementation of listener events for a pool-worker.
```
Or maybe
```suggestion
 * Basic interface describing the minimum required implementation of listener events for a pool worker.
```

File: src/pools/abstract-pool.ts
Comment (ST-DDT) [inline src/pools/abstract-pool.ts:62]:
```suggestion
 * Base class containing some shared logic for all poolifier pools.
```

File: src/pools/abstract-pool.ts
Comment (ST-DDT) [inline src/pools/abstract-pool.ts:74]:
```suggestion
   * List of currently active workers.
```
or
```suggestion
   * List of currently available workers.
```
?

File: src/pools/abstract-pool.ts
Comment (ST-DDT) [inline src/pools/abstract-pool.ts:230]:
```suggestion
   * Choose a worker for the next task.
   *
   * The default implementation uses a round robin algorithm to distribute the load.
```

File: src/pools/cluster/dynamic.ts
Comment (ST-DDT) [inline src/pools/cluster/dynamic.ts:42]:
```suggestion
   * Choose a worker for the next task.
   *
   * It will first check for and return an idle worker.
   * If all workers are busy, then it will try to create a new one up to the `max` worker count.
   * If the max worker count is reached, the emitter will emit a `FullPool` event and it will fall back to using a round robin algorithm to distribute the load.
```

File: src/pools/pool.ts
Comment (ST-DDT) [inline src/pools/pool.ts:9]: What happens to the tasks that are currently running/pending to run.

File: src/pools/thread/dynamic.ts
Comment (ST-DDT) [inline src/pools/thread/dynamic.ts:42]:
```suggestion
   * Choose a thread for the next task.
   *
   * It will first check for and return an idle thread.
   * If all threads are busy, then it will try to create a new one up to the `max` thread count.
   * If the max thread count is reached, the emitter will emit a `FullPool` event and it will fall back to using a round robin algorithm to distribute the load.
```

File: src/worker/abstract-worker.ts
Comment (ST-DDT) [inline src/worker/abstract-worker.ts:6]:
```suggestion
 * Base class containing some shared logic for all poolifier workers.
```

File: src/worker/abstract-worker.ts
Comment (ST-DDT) [inline src/worker/abstract-worker.ts:18]:
```suggestion
   * The maximum time to keep this worker alive while idle. The pool automatically checks and terminates this worker when the time expires.
```

File: src/worker/abstract-worker.ts
Comment (ST-DDT) [inline src/worker/abstract-worker.ts:26]: Last task queued or last time finished processing a task?

File: src/pools/abstract-pool.ts
Comment (jerome-benoit) [reply src/pools/abstract-pool.ts:101]: Since it's not a public, you can.

**Review by ST-DDT — COMMENTED** [review]
I am not sure about some of the comments, so take it with a grain of salt.

File: src/pools/abstract-pool.ts
Comment (jerome-benoit) [reply src/pools/abstract-pool.ts:81]: You can but keep a `get id() { return this messageId }`

Comment (Shinigami92) [reply src/pools/abstract-pool.ts:81]: @pioardi Yeah we could do that for backward compatibility reasons
But it as we already changed so much for the next major release, I think it's not necessary

File: src/pools/pool.ts
Comment (Shinigami92) [reply src/pools/pool.ts:9]: Yeah ... I think they just get killed, whatever they are doing.
@pioardi if this is the case, we should create an issue feature-request for providing/selecting a strategy/option whether to kill it with :fire: or to shutdown it gracefully and wait until the worker has finished its work

File: src/worker/abstract-worker.ts
Comment (Shinigami92) [reply src/worker/abstract-worker.ts:26]: The code says: Last time a message was send to the main worker :thinking:  
@pioardi Could you help here? Maybe you know your project better :slightly_smiling_face:

File: src/pools/abstract-pool.ts
Comment (Shinigami92) [reply src/pools/abstract-pool.ts:261]: #134

File: src/pools/abstract-pool.ts
Comment (Shinigami92) [reply src/pools/abstract-pool.ts:101]: #134

File: src/pools/abstract-pool.ts
Comment (Shinigami92) [reply src/pools/abstract-pool.ts:81]: #134

**Comment by Shinigami92** [comment]
@pioardi Good morning :slightly_smiling_face:  

Today I want to tackle PR 134, maybe 133 and also maybe 105 (maybe the last two jerome wants to work on?)
I also want to try an idea that came into my mind early morning today... just need to find out if its good or not

So are you willing to approve this PR now, so I can move on in other PRs based on an updated master? That could reduce branch conflicts
Most of the related improvement stuff I already moved into new issues

File: src/pools/abstract-pool.ts
Comment (pioardi) [reply src/pools/abstract-pool.ts:84]: As is, is fine for me

File: src/pools/abstract-pool.ts
Comment (pioardi) [reply src/pools/abstract-pool.ts:94]: It is ok for now

File: src/pools/pool.ts
Comment (pioardi) [reply src/pools/pool.ts:9]: There is already an open enhancement for that [#70](https://github.com/pioardi/poolifier/issues/70), but we need to think carefully about this since that could be in conflict with the maxInactiveTime and slow down performances also.
The map "tasks" into our pool implementations could help on that.

File: src/worker/abstract-worker.ts
Comment (pioardi) [reply src/worker/abstract-worker.ts:26]: This property indicate when the last task was processed by this worker.
If you submit a task at 11:00:02 and the tasks finish at 11:00:04, this variable value will be 11:00:04 for the worker that worked on the task

**Comment by Shinigami92** [comment]
@pioardi You need to click on the approve button, if you forgot that ^^
Or is there something I need to do additionally in this PR?

**Comment by pioardi** [comment]
I think we are ready, it is a great improvement :D

**Review by pioardi — APPROVED** [review]
Approved

**Pull request merged**
